direct/algorithms/mri.py

```diff
@@ -0,0 +1,179 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+
+from typing import Any, Callable, Dict
+
+import numpy as np
+import torch
+from torch import nn
+
+from direct.algorithms.optimization import MaximumEigenvaluePowerMethod
+from direct.data.transforms import view_as_complex, view_as_real
+from direct.utils import DirectTransform
+
+
+class EspiritCalibration(DirectTransform):
+    """Estimates sensitivity maps estimated with the ESPIRIT calibration method as described in [1]_.
+
+    We adapted code for ESPIRIT method adapted from [2]_.
+
+    References
+    ----------
+
+    .. [1] Uecker M, Lai P, Murphy MJ, Virtue P, Elad M, Pauly JM, Vasanawala SS, Lustig M. ESPIRiT--an eigenvalue
+        approach to autocalibrating parallel MRI: where SENSE meets GRAPPA. Magn Reson Med. 2014 Mar;71(3):990-1001.
+        doi: 10.1002/mrm.24751. PMID: 23649942; PMCID: PMC4142121.
+    .. [2] https://github.com/mikgroup/sigpy/blob/1817ff849d34d7cbbbcb503a1b310e7d8f95c242/sigpy/mri/app.py#L388-L491
+
+    """
+
+    def __init__(
+        self,
+        backward_operator: Callable,
+        threshold: float = 0.05,
+        kernel_size: int = 6,
+        crop: float = 0.95,
+        max_iter: int = 100,
+        kspace_key: str = "masked_kspace",
+    ):
+        """Inits :class:`EstimateSensitivityMap`.
+
+        Parameters
+        ----------
+        backward_operator: Callable
+            The backward operator, e.g. some form of inverse FFT (centered or uncentered).
+        threshold: float, optional
+            Threshold for the calibration matrix. Default: 0.05.
+        kernel_size: int, optional
+            Kernel size for the calibration matrix. Default: 6.
+        crop: float, optional
+            Output eigenvalue cropping threshold. Default: 0.95.
+        max_iter: int, optional
+            Power method iterations. Default: 30.
+        kspace_key: str
+            K-space key. Default `masked_kspace`.
+        """
+        self.backward_operator = backward_operator
+        self.threshold = threshold
+        self.kernel_size = kernel_size
+        self.crop = crop
+        self.max_iter = max_iter
+        self.kspace_key = kspace_key
+        super().__init__()
+
+    @staticmethod
+    def crop_to_acs(acs_mask: torch.Tensor, kspace: torch.Tensor) -> torch.Tensor:
+        """Crops k-space to autocalibration region given the acs_mask.
+
+        Parameters
+        ----------
+        acs_mask : torch.Tensor
+            Autocalibration mask.
+        kspace : torch.Tensor
+            K-space.
+
+        Returns
+        -------
+        torch.Tensor
+            Cropped k-space.
+        """
+        nonzero_idxs = torch.nonzero(acs_mask)
+        x, y = nonzero_idxs[..., 0], nonzero_idxs[..., 1]
+        xl, xr = x.min(), x.max()
+        yl, yr = y.min(), y.max()
+        return kspace[:, xl : xr + 1, yl : yr + 1]
```

Comment:
Don't we have this somewhere already? If so, wouldn't it make sense to have this as a separate function?

Reply:
Done